enable static graph training in DDP

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55248 [WIP] enable static graph training in DDP**

This PR provides enable static graph training when users call _set_static_graph(). This can help support more use cases in DDP without performance regression, also can potentially improve performance when there are unused parameters in the graph.
1. first iteration records graph states like how many times a grad is calculated, whether the grad is used or not. then first iteration queues a delay_all_reduce call back to all reduce grads.
2. Since autograd call back is associated with current target graph task, the delay_all_all call back should be associated with out-most backward graph task. A DDP sink layer is added in DDP forward loop so that we can queue the delay_all_reduce call back in the sink layer.
3. after first iterations, DDP will use the saved graph states to determine whether a grad is used or not. whether a grad is ready for communication.
4. rebuilt bucket is called in second iteration, after graph states are recorded in first iteration.
5. if the graph states change, DDP will throw errors

Differential Revision: [D27539964](https://our.internmc.facebook.com/intern/diff/D27539964/)

zhaojuanmao committed:
[WIP] enable static graph training in DDP

This PR provides enable static graph training when users call _set_static_graph(). This can help support more use cases in DDP without performance regression, also can potentially improve performance when there are unused parameters in the graph.
1. first iteration records graph states like how many times a grad is calculated, whether the grad is used or not. then first iteration queues a delay_all_reduce call back to all reduce grads.
2. Since autograd call back is associated with current target graph task, the delay_all_all call back should be associated with out-most backward graph task. A DDP sink layer is added in DDP forward loop so that we can queue the delay_all_reduce call back in the sink layer.
3. after first iterations, DDP will use the saved graph states to determine whether a grad is used or not. whether a grad is ready for communication.
4. rebuilt bucket is called in second iteration, after graph states are recorded in first iteration.
5. if the graph states change, DDP will throw errors

Differential Revision: [D27539964](https://our.internmc.facebook.com/intern/diff/D27539964/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/reducer.hpp:
```diff
@@ -96,7 +118,8 @@ class Reducer {
   inline bool should_rebuild_buckets() const {
-    return !find_unused_parameters_ && !has_rebuilt_bucket_;
+    return !find_unused_parameters_ && !has_rebuilt_bucket_
+      && (unused_parameters_.size() == 0);
```
Why do we need this additional check for `unused_parameters`? Is there a scenario where `unused_parameters` will be non-zero despite `find_unused_parameters=True`?

zhaojuanmao replied:
yes, if users set find_unused_parameters=false and only set_static_graph(), it is possible unused_parameters_ size is non zero

pritamdamania87 commented on torch/nn/parallel/distributed.py:
```diff
@@ -108,6 +109,19 @@ class _DDPUnevenInputsConfig(NamedTuple):
+
+    @staticmethod
+    def backward(ctx, input):
+        Variable._execution_engine.queue_callback(ctx.reducer._delay_all_reduce)
```
Why do we need a DDPSink function like this? Can't all the delay_allreduce logic go into `finalize_backward`?

zhaojuanmao replied:
right now finalize_backward is queued when all gradients are completed. 

1) at first iteration, we do not know when the gradients is completed, so we can not queue in the same way as how finalize_backward is queued. 2) we can queue the callback before backward starts, but it needs to be associated with the out most backward task

rohan-varma replied:
I think the current version of this PR works well, but as an enhancement, we could consider:

1) Don't wait to enqueue finalize_backward until all buckets have allreduce launched, instead, enqueue it immediately and refactor it to block waiting for allreduces to be kicked off (an idea @pritamdamania87 mentioned)
2) Move enqueueing to this sink function
3) For static graph, we'd then queue either delay allreduce or regular finalize_backward depending on the business logic.

This has the benefit that there's only a single well-defined place where the callback is being enqueued.

zhaojuanmao replied:
move queueing finalize_backward to backward end should work, but we should be careful about it. e..g, not forgot to checking all buckets are computed and etc

pritamdamania87 commented on test/distributed/test_c10d.py:
```diff
@@ -4218,100 +4218,121 @@ def forward(self, inp):
             return x
 
-    def _test_ddp_checkpointing(self, checkpoint_once, process_group, use_bucket_view, find_unused_parameters=False):
+    def _prepare_dummy_data(self):
```
Can you also enable all the tests in https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/distributed/pipe_with_ddp_test.py and verify they work with `set_static_graph`? Can be done in a separate PR.

zhaojuanmao replied:
sounds good, I will add verify them in follow up PR

rohan-varma reviewed (commented):
Are we planning to ensure this API works with `DDP._set_params_and_buffers_to_ignore_for_model` as well, or is support for that not planned in the first iteration of this feature?

SciPioneer commented:
I wonder if we can simplify DDP implementation and avoid creating `delay_all_reduce` mechanism as well as `DDPSink` here.

If we need to record the graph states or unused parameters in the first iteration, this can be easily done in a few lines in a communication hook (collecting unused parameters while all-reduce rather than deferring all-reduce). These graph states can be provided as a comm hook state, which is also a new field of DDP model.

Therefore, we can do the following instead:
1) Provide a built-in comm hook that collects graph states in the first iteration and does normal-allreduce. The user needs to feed a graph state dictionary (which can be a new data member of DDP model) to this hook.
2) Rebuild the buckets based on the graph state dictionary if any.

In terms of API, a user can call `model.register_comm_hook(state=model.get_graph_states(), hook=static_graph_hook)` instead of `model.set_static_graph()`. This can improve the composability of DDP.

zhaojuanmao commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -455,31 +520,38 @@ void Reducer::autograd_hook(VariableIndex index) {
+  // If it is static graph, after 1st iteration, check a avariable
+  // is ready for communication based on numGradHooksTriggeredMap_.
+  if (static_graph_ && num_iterations_ > 1) {
+    TORCH_CHECK(
+        numGradHooksTriggeredMapPerIteration_[index] > 0,
+        "Your training graph is changed in this iteration, "
+        "try not to specify static_graph_ to be true in this case.");
```
if the number never went to 0, it will not be marked as Ready for communication, then it will not trigger finalize_backward, and ensure_prior_reduction_finished() will trow errors in this case

zhaojuanmao commented:
> Are we planning to ensure this API works with `DDP._set_params_and_buffers_to_ignore_for_model` as well, or is support for that not planned in the first iteration of this feature?

I just checked how `DDP._set_params_and_buffers_to_ignore_for_model` is implemented, if I understand it correctly, '_set_params_and_buffers_to_ignore_for_model' filters ignored parameters in python construction time, right? if so, set_static_graph() API should be compatible with it

zhaojuanmao commented:
> I wonder if we can simplify DDP implementation and avoid creating `delay_all_reduce` mechanism as well as `DDPSink` here.
> 
> If we need to record the graph states or unused parameters in the first iteration, this can be easily done in a few lines in a communication hook (collecting unused parameters while all-reduce rather than deferring all-reduce). These graph states can be provided as a comm hook state, which is also a new field of DDP model.
> 
> Therefore, we can do the following instead:
> 
> 1. Provide a built-in comm hook that collects graph states in the first iteration and does normal-allreduce. The user needs to feed a graph state dictionary (which can be a new data member of DDP model) to this hook.
> 2. Rebuild the buckets based on the graph state dictionary if any.
> 
> In terms of API, users can call `model.register_comm_hook(state=model.get_graph_states(), hook=static_graph_hook)` instead of `model.set_static_graph()`. This can improve the composability of DDP.

The states is not only to identify unused_parameters, but also how many times a grad could be calculated in a backward pass. e.g., when there are reentrant backwards, grads could be calculated multiple times. 
Without knowing this state, we can not know whether grad is completed all calculation and is ready for all reduce, right? 
That is why we need delay all reduce in the first iteration. 

by the way, most of codes here, like delay all reduce, will be reused for _delay_all_reduce() API in the near future. with _delay_all_reduce() API, users can skip DDP reducer bucketing optimization for some edge cases that current DDP can not support

rohan-varma commented:
> > Are we planning to ensure this API works with `DDP._set_params_and_buffers_to_ignore_for_model` as well, or is support for that not planned in the first iteration of this feature?
> 
> I just checked how `DDP._set_params_and_buffers_to_ignore_for_model` is implemented, if I understand it correctly, '_set_params_and_buffers_to_ignore_for_model' filters ignored parameters in python construction time, right? if so, set_static_graph() API should be compatible with it

Yes, my thinking is it should be compatible too. Feel free to add tests if it seems useful, it could help ensuring we have full feature parity with this new fature.

SciPioneer commented:
> The states is not only to identify unused_parameters, but also how many times a grad could be calculated in a backward pass. e.g., when there are reentrant backwards, grads could be calculated multiple times.

A follow-up question: Is "how many times a grad could be calculated in a backward pass" = "how many times a grad is all-reduced"?

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -108,6 +109,19 @@ class _DDPUnevenInputsConfig(NamedTuple):
+# Add a DDPSink to queue call back of out-most backward/graph task,
+# this helps call back is fired after all gradients' calculation
+# is completed.
+class DDPSink(Function):
```
Can we test this change when DDP model outputs a tuple (a, b) of tensors? I'm afraid we may hit the issue in https://github.com/pytorch/pytorch/issues/55509, but I hope I'm wrong.

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4200,8 +4200,8 @@ def test_ddp_weight_sharing(self):
             super().__init__()
-            self.l1 = nn.Linear(2000, 2000)
-            self.l2 = nn.Linear(2000, 2000)
+            self.l1 = nn.Linear(20, 20)
```
Wondering why we decrease the model size here?

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4218,100 +4218,121 @@ def forward(self, inp):
+        ddp_bs = 16
+        bs = ddp_bs * self.world_size
+        input = torch.rand((bs, 20), device="cuda", requires_grad=True)
+        target = torch.randn((bs, 20), device="cuda")
```
Maybe use `torch.cuda.current_device()` to ensure the tensors are on the right device?

zhaojuanmao replied:
torch.cuda.set_device(self.rank) was set at the beginning

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4218,100 +4218,121 @@ def forward(self, inp):
-        output_tensor = model(input_tensor)
-        output_tensor.sum().backward()
-        return model
+        model = copy.deepcopy(input_model).cuda()
```
Seems like we don't need this deepcopy (can keep the next one)?

zhaojuanmao replied:
we need this one, this is for local_model, next one is for ddp_model, we need to compare their grads

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4218,100 +4218,121 @@ def forward(self, inp):
+        input, ddp_input, target, ddp_target = self._prepare_dummy_data()
+        loss = nn.MSELoss()
+        for i in range(5):
+            model.zero_grad(set_to_none=False)
```
What's default arg for `zero_grad`? Do users need to explicitly add `set_to_none=False` if they are using static graph training?

zhaojuanmao replied:
in default, set_to_none=False, we do not have to set it explicitly

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4218,100 +4218,121 @@ def forward(self, inp):
+                                                 process_group=process_group,
+                                                 use_bucket_view=use_bucket_view,
+                                                 find_unused_parameters=True,
+                                                 static_graph=True)
```
Should we keep the previous test that also tested with no static graph? 

And for the error, can we add string to recommend `use_static_graph` if it works for their use case to error message?

zhaojuanmao replied:
sounds good!

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -137,6 +137,8 @@ Reducer::Reducer(
           gradAccToVariableMap_[grad_accumulator.get()] = index;
         }
 
+        numGradHooksTriggeredMap_[index] = 0;
```
I am adding some additional maps in https://github.com/pytorch/pytorch/pull/55075 as well. I wonder if we should eventually make some sort of struct to hold all of these so that they are easy to update all together and not scattered all over the place in `reducer.cpp`.

zhaojuanmao replied:
possibly we can do later on when more semantically group-able fields are added

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -429,6 +431,69 @@ void Reducer::set_divide_factor() {
+void Reducer::delay_all_reduce() {
+  std::lock_guard<std::mutex> lock(this->mutex_);
+
+  if (!static_graph_ || num_iterations_ > 1) {
```
Could we add this check in Python? Advantages are removing the call to C++ and making it clearer to users that it's only called once in static graph training.

zhaojuanmao replied:
I feel this is part of reducer internals, it is good to keep it in C++. otherwise we need to expose num_iterations_ to python as well

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -429,6 +431,69 @@ void Reducer::set_divide_factor() {
+  if (should_collect_runtime_stats()) {
+    record_backward_compute_end_time();
+  }
+  // The autograd engine uses the default stream when running callbacks, so we
```
Can we put this into a helper function since it's also used around `finalize_backward`?

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -429,6 +431,69 @@ void Reducer::set_divide_factor() {
+
+  // prepare to set unused_parameters_, if it is static graph,
+  // unused_parameters_ will not change after 1st iteration.
+  if (static_graph_ && num_iterations_ == 1) {
```
`static_graph_` is always true after the check at the beginning, can we remove it here and below?

zhaojuanmao replied:
I am thinking of this function could be used for _delay_all_reduce run mode in the future. but for now I can remove it for easier understanding

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -429,6 +431,69 @@ void Reducer::set_divide_factor() {
+         variable_index++) {
+      const auto index = VariableIndex(replica_index, variable_index);
+      // set unused_parameters_
+      if (static_graph_ && num_iterations_ == 1 && numGradHooksTriggeredMap_[index] == 0) {
```
nit: It might be easier to index with `size_t variable_index` over `VariableIndex` structure since `replica_index` is always 0, will help the code be simpler

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -455,31 +520,38 @@ void Reducer::autograd_hook(VariableIndex index) {
+  // is ready for communication based on numGradHooksTriggeredMap_.
+  if (static_graph_ && num_iterations_ > 1) {
+    TORCH_CHECK(
+        numGradHooksTriggeredMapPerIteration_[index] > 0,
```
This checks > 0, but if it is computed as triggered N times in the first iteration, but then is computed K < N times later on, how do we guard against this?

rohan-varma replied:
Nevermind, IIUC we will call it N times and it will go to negative, throwing this error

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -544,6 +627,14 @@ void Reducer::mark_variable_ready(VariableIndex index) {
+    if (static_graph_) {
+      TORCH_CHECK(
+        false,
+        "Your training graph is changed in this iteration, "
```
nit: combine these into a constant str and use that everywhere it is repeated?

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -522,6 +594,17 @@ void Reducer::all_reduce_local_used_map() {
+  // rebuilt based on rebuilt_params_ and rebuilt_param_indices_, and then
+  // will be broadcasted and initialized. Also we only need to dump tensors
+  // and parameter indices of one replica.
+  push_rebuilt_params(index);
```
Wondering why we moved this from `autograd_hook` to mark_variable_ready?

zhaojuanmao replied:
if grads are triggered more than once, this index will be put into rebuilt_params_ multiple times.  we only need to record this index once when this grad is really ready for communication.

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -544,6 +627,14 @@ void Reducer::mark_variable_ready(VariableIndex index) {
   // Something is wrong if all variables contained in this bucket replica have
   // already been marked as ready.
   if (replica.pending == 0) {
+    // For static graph, it should not get here. If it does, mostly
```
Wondering Is there a concrete example of how the application can reach this state with static_graph training (graph is not static, but concretely what does the model end up doing?), and what is meant by mostly?

rohan-varma commented on torch/lib/c10d/reducer.hpp:
```diff
@@ -25,6 +25,28 @@ constexpr int kDefaultBucketBytesCap = int(25 * 1024 * 1024);
+    variable_index = variable_index_;
+  }
+
+  static size_t hash(const VariableIndex& key) {
```
We can probably remove this if we make the data structures to be indexed only by replica_index

rohan-varma commented on torch/lib/c10d/reducer.hpp:
```diff
@@ -421,6 +435,11 @@ class Reducer {
 
   bool static_graph_ = false;
 
+  // Map will not change after 1st iteration.
```
nit: Would be good to add a line explaining what the maps are responsible for doing.

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -108,6 +109,19 @@ class _DDPUnevenInputsConfig(NamedTuple):
+
+    @staticmethod
+    def backward(ctx, input):
+        Variable._execution_engine.queue_callback(ctx.reducer._delay_all_reduce)
```
Should we keep track of iteration here and gate it in python itself?

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -4218,100 +4218,121 @@ def forward(self, inp):
+                                         run_checkpoint=True)
+
+    # TODO adding more tests
+    def test_more(self):
```
Would be good to have a test where we validate the appropriate errors are raised if the graph doesn't end up being static at some iteration.

zhaojuanmao replied:
yeah, sounds good

rohan-varma reviewed (commented):
Looks great overall, thanks for adding this! Left some comments inline, will accept after more tests are added per your todo

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -455,31 +520,38 @@ void Reducer::autograd_hook(VariableIndex index) {
-  push_rebuilt_params(index);
+  if (static_graph_ && num_iterations_ == 1) {
+    numGradHooksTriggeredMap_[index] += 1;
+    return;
```
IIUC this is what prevents buckets from being marked as ready and then go to allreduce in the first iteration for static graph training. Hopefully it's not too prone to breaking via a refactor or anything like that. Maybe it's worth it to add `TORCH_CHECK(!static_graph_training_ || num_iterations > 1)` below to ensure we don't get to the actual mark variable ready logic when we should be running delay allreduce?

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -455,31 +520,38 @@ void Reducer::autograd_hook(VariableIndex index) {
   // discovered in the `prepare_for_backward` function and their indexes stored
   // in the `unused_parameters_` vector.
-  if (!has_marked_unused_parameters_ && find_unused_parameters_) {
+  if (!has_marked_unused_parameters_) {
```
Won't removing the `&& find_unused_parameters_` check here change things for regular non static graph training? 

For example if !static_graph and find_unused_parameters=False, this will call mark_variable_ready on each unused index, although I guess it should be a no-op since that list should be empty for this case?

zhaojuanmao replied:
yes, for the case '!static_graph and find_unused_parameters=False', unused_parameters_ is empty and thus it is no-op. 

 I removed this check, because in static graph training mode, 'unused_parameters_' is always detected  no matter find_unused_parameters is true or false.

rohan-varma commented:
Would also be great to add a test where graph is static but different across ranks to ensure we can cover that as well

xw285cornell commented:
This is great! Please make sure _set_params_and_buffers_to_ignore_for_model works also :)

xw285cornell commented:
@zhaojuanmao qq: will we be able to rebuild bucket at the 2nd iteration, if we do the static_graph way, with some unused parameters in the graph? (the unused param will not change x-iterations).

zhaojuanmao commented:
> @zhaojuanmao qq: will we be able to rebuild bucket at the 2nd iteration, if we do the static_graph way, with some unused parameters in the graph? (the unused param will not change x-iterations).

Yes, we can do that

zhaojuanmao committed:
Update on "[WIP] enable static graph training in DDP"


This PR provides enable static graph training when users call _set_static_graph(). This can help support more use cases in DDP without performance regression, also can potentially improve performance when there are unused parameters in the graph.
1. first iteration records graph states like how many times a grad is calculated, whether the grad is used or not. then first iteration queues a delay_all_reduce call back to all reduce grads.
2. Since autograd call back is associated with current target graph task, the delay_all_all call back should be associated with out-most backward graph task. A DDP sink layer is added in DDP forward loop so that we can queue the delay_all_reduce call back in the sink layer.
3. after first iterations, DDP will use the saved graph states to determine whether a grad is used or not. whether a grad is ready for communication.
4. rebuilt bucket is called in second iteration, after graph states are recorded in first iteration.
5. if the graph states change, DDP will throw errors

Differential Revision: [D27539964](https://our.internmc.facebook.com/intern/diff/D27539964/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[WIP] enable static graph training in DDP"


This PR provides enable static graph training when users call _set_static_graph(). This can help support more use cases in DDP without performance regression, also can potentially improve performance when there are unused parameters in the graph.
1. first iteration records graph states like how many times a grad is calculated, whether the grad is used or not. then first iteration queues a delay_all_reduce call back to all reduce grads.
2. Since autograd call back is associated with current target graph task, the delay_all_all call back should be associated with out-most backward graph task. A DDP sink layer is added in DDP forward loop so that we can queue the delay_all_reduce call back in the sink layer.
3. after first iterations, DDP will use the saved graph states to determine whether a grad is used or not. whether a grad is ready for communication.
4. rebuilt bucket is called in second iteration, after graph states are recorded in first iteration.
5. if the graph states change, DDP will throw errors

Differential Revision: [D27539964](https://our.internmc.facebook.com/intern/diff/D27539964/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[WIP] enable static graph training in DDP"


This PR provides enable static graph training when users call _set_static_graph(). This can help support more use cases in DDP without performance regression, also can potentially improve performance when there are unused parameters in the graph.
1. first iteration records graph states like how many times a grad is calculated, whether the grad is used or not. then first iteration queues a delay_all_reduce call back to all reduce grads.
2. Since autograd call back is associated with current target graph task, the delay_all_all call back should be associated with out-most backward graph task. A DDP sink layer is added in DDP forward loop so that we can queue the delay_all_reduce call back in the sink layer.
3. after first iterations, DDP will use the saved graph states to determine whether a grad is used or not. whether a grad is ready for communication.
4. rebuilt bucket is called in second iteration, after graph states are recorded in first iteration.
5. if the graph states change, DDP will throw errors

Differential Revision: [D27539964](https://our.internmc.facebook.com/intern/diff/D27539964/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[WIP] enable static graph training in DDP"


This PR provides enable static graph training when users call _set_static_graph(). This can help support more use cases in DDP without performance regression, also can potentially improve performance when there are unused parameters in the graph.
1. first iteration records graph states like how many times a grad is calculated, whether the grad is used or not. then first iteration queues a delay_all_reduce call back to all reduce grads.
2. Since autograd call back is associated with current target graph task, the delay_all_all call back should be associated with out-most backward graph task. A DDP sink layer is added in DDP forward loop so that we can queue the delay_all_reduce call back in the sink layer.
3. after first iterations, DDP will use the saved graph states to determine whether a grad is used or not. whether a grad is ready for communication.
4. rebuilt bucket is called in second iteration, after graph states are recorded in first iteration.
5. if the graph states change, DDP will throw errors

Differential Revision: [D27539964](https://our.internmc.facebook.com/intern/diff/D27539964/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[WIP] enable static graph training in DDP"


This PR provides enable static graph training when users call _set_static_graph(). This can help support more use cases in DDP without performance regression, also can potentially improve performance when there are unused parameters in the graph.
1. first iteration records graph states like how many times a grad is calculated, whether the grad is used or not. then first iteration queues a delay_all_reduce call back to all reduce grads.
2. Since autograd call back is associated with current target graph task, the delay_all_all call back should be associated with out-most backward graph task. A DDP sink layer is added in DDP forward loop so that we can queue the delay_all_reduce call back in the sink layer.
3. after first iterations, DDP will use the saved graph states to determine whether a grad is used or not. whether a grad is ready for communication.
4. rebuilt bucket is called in second iteration, after graph states are recorded in first iteration.
5. if the graph states change, DDP will throw errors

Differential Revision: [D27539964](https://our.internmc.facebook.com/intern/diff/D27539964/)

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -571,13 +651,17 @@ void Reducer::checkAndRaiseMarkedTwiceError(size_t curVariableIndex) {
         "Please make sure model parameters are not shared across multiple ",
-        "concurrent forward-backward passes",
+        "concurrent forward-backward passes. or try to use _set_static_graph() ",
+        "as a workaround if this module parameter does not change ",
```
nit: `if the module graph does not change`?

rohan-varma commented on torch/lib/c10d/reducer.cpp:
```diff
@@ -594,12 +678,27 @@ void Reducer::checkAndRaiseMarkedTwiceError(size_t curVariableIndex) {
-        "`torch.nn.parallel.DistributedDataParallel`.");
+        "`torch.nn.parallel.DistributedDataParallel`. If unused parameters ",
+        "in the model do not change over iterations, You can try to use ",
+        "_set_static_graph() as a workaround.");
```
also specify that the rest of the graph needs to be static?

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -567,6 +583,8 @@ def __setstate__(self, state):
         # Builds reducer
         self._ddp_init_helper(parameters, expect_sparse_gradient, param_to_name_mapping)
+        if self.static_graph:
+            self._set_static_graph()
```
if `find_unused_parameters=True` is set in ddp, should we add a warnings.warn once indicating that the flag will have no effect with static graph training?

rohan-varma replied:
Just realized that this path runs when ddp model is saved/loaded. We can add a warning here as well as in the regular init path

rohan-varma commented on torch/lib/c10d/reducer.hpp:
```diff
@@ -457,6 +486,7 @@ class Reducer {
   // Raises appropriate error if mark_variable_ready is called on the same
   // variable twice, which is unexpected.
   void checkAndRaiseMarkedTwiceError(size_t curVariableIndex);
+
```
nit: extra line

rohan-varma reviewed (approved):
Awesome, looks great! Thanks for working through all the cases and adding comprehensive tests to make sure this works with all different features. Have some very minor comments which can be addressed here or in a follow up as needed.

zhaojuanmao committed:
Update on "[WIP] enable static graph training in DDP"


This PR provides enable static graph training when users call _set_static_graph(). This can help support more use cases in DDP without performance regression, also can potentially improve performance when there are unused parameters in the graph.
1. first iteration records graph states like how many times a grad is calculated, whether the grad is used or not. then first iteration queues a delay_all_reduce call back to all reduce grads.
2. Since autograd call back is associated with current target graph task, the delay_all_all call back should be associated with out-most backward graph task. A DDP sink layer is added in DDP forward loop so that we can queue the delay_all_reduce call back in the sink layer.
3. after first iterations, DDP will use the saved graph states to determine whether a grad is used or not. whether a grad is ready for communication.
4. rebuilt bucket is called in second iteration, after graph states are recorded in first iteration.
5. if the graph states change, DDP will throw errors

Differential Revision: [D27539964](https://our.internmc.facebook.com/intern/diff/D27539964/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "[WIP] enable static graph training in DDP"


This PR provides enable static graph training when users call _set_static_graph(). This can help support more use cases in DDP without performance regression, also can potentially improve performance when there are unused parameters in the graph.
1. first iteration records graph states like how many times a grad is calculated, whether the grad is used or not. then first iteration queues a delay_all_reduce call back to all reduce grads.
2. Since autograd call back is associated with current target graph task, the delay_all_all call back should be associated with out-most backward graph task. A DDP sink layer is added in DDP forward loop so that we can queue the delay_all_reduce call back in the sink layer.
3. after first iterations, DDP will use the saved graph states to determine whether a grad is used or not. whether a grad is ready for communication.
4. rebuilt bucket is called in second iteration, after graph states are recorded in first iteration.
5. if the graph states change, DDP will throw errors

Differential Revision: [D27539964](https://our.internmc.facebook.com/intern/diff/D27539964/)

[ghstack-poisoned]
